### ISSUE TITLE
[data grid] Regression, new different behaviour on mobile leads to hidden column header labels

### ISSUE THREAD

**Issue by Reckless-Satoshi** [mui/mui-x#9776]
### Steps to reproduce 🕹

On Mobile, column headers labels are hidden by default by the sorting and menu buttons. This was not the behavior a few versions back (I cannot pinpoint exactly when this regression took place, or if it's a problem inherited form a dependency):
![a](https://github.com/mui/mui-x/assets/90936742/1c448b65-93fc-48eb-810c-2eeaa94f3aa8)
(Note how `username` is not visible)

On desktop:
![b](https://github.com/mui/mui-x/assets/90936742/512f5a2e-004b-48bc-848f-73818bab3a79)

**Link to live example:**
Any Data Grid in https://mui.com/x/react-data-grid with a column label that fits tight in the header has this "issue". I used as example the first DataGrid in https://mui.com/x/react-data-grid/column-definition/ . Just press F12 and enable mobile view to see how annoying is not being able to read headers that should otherwise fit in.

### Current behavior 😯

Since some recent MUI-X update, the mobile view of DataGrid will always show the Sort and Menu (triple dot) buttons in the column headers. This leads to the labels losing visibility. Early v6 versions of MUI-X DataGrid would only show the sort and menu buttons when the header is touched in mobile or hovered in desktop. As my app mobile DataGrid relied on very tight header labels in order to optimize use of screen space, now they are unreadable by default after updating to latest MUI / MUI-X.

### Expected behavior 🤔

The column header should remain readable unless hovered or tapped.

### Context 🔦

I have tried to replicate the old behaviour by dynamically turning on/off  `disableColumnMenu` and `Sortable` on the `columns` definitions with the callbacks `onColumnHeaderEnter` and `onColumnHeaderOut`. However the re-rendering with new `columns` results in a jarring jittering of the whole DataGrid.  

Any workaround or help is appreciated. The only option I contemplate at the moment is to permanently disable column sorting and menu to recover readability at the expense of functionality.

**Comment by Reckless-Satoshi** [comment]
@m4theushw very useful!

I have been able to _almost_ replicate the old behaviour (show header icons on tap) with
```
const StyledDataGrid = styled(DataGrid)({
  '@media (hover: none)': {
    '&& .MuiDataGrid-menuIcon': {
      width: 0,
      visibility: 'hidden',
    },
    '&& .MuiDataGrid-sortIcon': {
      width: 0,
      visibility: 'hidden',
    }
  },
  '&& .MuiDataGrid-columnHeader--sorted .MuiDataGrid-menuIcon': {
    width: 'auto',
    visibility: 'visible',
  },
  '&& .MuiDataGrid-columnHeader--sorted .MuiDataGrid-sortIcon': {
    width: 'auto',
    visibility: 'visible',
  }
});
```
The column labels are were tightest do still lose 2 or 3 characters, but it is more readable now.

**Desktop**
![desktop](https://github.com/mui/mui-x/assets/90936742/edc70530-4523-485e-931d-376555fc6799)
**Mobile**
![mobile](https://github.com/mui/mui-x/assets/90936742/bdd5e8bd-fb53-4211-bbfb-3ff4ee7907d5)
_Mobile and Desktop used to look identical_
**Mobile with CSS fix above**
![with_fix](https://github.com/mui/mui-x/assets/90936742/6f4d797b-bf4f-451d-b4fb-1be010face15)

**Comment by MBilalShafi** [comment]
@mui/xgrid, this is a corner case on mobile screens where below a specific column width, the text seems to disappear, to me, we could trade-off this with the benefit we achieve by showing the hover icons by default on mobile and prevent multiple taps/confusing UX.

Any opinions will be appreciated!

CC @gerdadesign

**Comment by cherniavskii** [comment]
@MBilalShafi

Maybe we could improve column header truncation on mobile?
There's plenty of space wasted because we render sorting and filter icons even when they're visually hidden:

![](https://user-images.githubusercontent.com/90936742/255691516-bdd5e8bd-fb53-4211-bbfb-3ff4ee7907d5.png)

**Comment by MBilalShafi** [comment]
> Maybe we could improve column header truncation on mobile?

True, this could certainly free up some space, especially when the icons have `visibility: hidden` but they do actually take up some space.

**Comment by KenanYusuf** [comment]
Fixed by https://github.com/mui/mui-x/pull/17375
